odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1471 @@
+
+        y : Element in domain of F^*.
+        """
+
+        if y is not None and not isinstance(y, LinearSpaceVector):
+            raise TypeError(
+                'vector {!r} not None or a LinearSpaceVector instance.'
+                ''.format(y))
+
+        if y not in convex_conj_f.domain:
+            raise TypeError(
+                'vector {} not in the domain of the functional {}.'
+                ''.format(y, convex_conj_f.domain))
+
+        super().__init__(domain=convex_conj_f.domain,
+                         linear=convex_conj_f.is_linear,
+                         smooth=convex_conj_f.is_smooth,
+                         concave=convex_conj_f.is_concave,
+                         convex=convex_conj_f.is_convex)
+
+        self._orig_convex_conj_f = convex_conj_f
+        self._y = y
+
+        # TODO:
+        # The Lipschitz constant for the gradient can be bounded, by using
+        # triangle inequality. However: is it the tightest bound?
+
+    def _call(self, x):
+        """Applies the functional to the given point.
+
+        Returns
+        -------
+        `self(x)` : `element` in the `field` of the ``domain``.
+            Evaluation of the functional.
+        """
+        return self._orig_convex_conj_f(x) + x.inner(self._y)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluated
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        return self._orig_convex_conj_f.gradient + ConstantOperator(self._y)
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the ConvexConjugateTranslation
+        functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator
+
+        Returns
+        -------
+        out : Operator
+            Domain and range equal to domain of functional
+        """
+        return proximal_quadratic_perturbation(
+            self._orig_convex_conj_f.proximal, a=0, u=self._y)(sigma)
+
+    # TODO: Should it be added?
+    # Note: This would only be valide when f is proper convex and lower-
+    # semincontinuous.
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of the functional."""
+        raise NotImplementedError
+
+
+class ConvexConjugateFuncScaling(Functional):
+    """ The ``Functional`` representing (scaling * F(.))^*.
+
+    Calculate the convex conjugate functional of the scaled function
+    sclaing * F(x).
+
+    This is calculated according to the rule
+
+        (scaling * F(.))^* (x) = scaling * F^*(x/scaling)
+
+    where ``scaling`` is the scaling parameter. Note that this does not allow
```

Comment:
Remove negations: `Note that this only allows for scaling with strictly positive values.`